### PR TITLE
LLT-5212: Use structured events in FFI event callback

### PR DESCRIPTION
**Pull request by mathiaspeters**
### Problem
Now that we have the type safety in our FFI layer that UniFFI provides there is no need to keep our FFI event callback dealing with strings. Natlab still relies on its own types and strings to work and changing that would be a big piece of work so for now, only a serialization function is created to keep natlab working. A task for miigrating natlab to the generated types has been create and  will be done later

### Solution
- Make UniFFI-related code know about the event structs
- Stop serializing the events to json before invoking the callback
- Serialize the events in python to keep natlab working

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean ([requirements](../blob/main/docs/git_commit_messages_requirements.md))
- [x] README.md is updated
- [x] Functionality is covered by unit or integration tests

### CODE REVIEW

File: src/ffi/types.rs [hunk 1]
```diff
@@ -7,15 +8,15 @@ use std::panic::RefUnwindSafe;
...
+    fn event(&self, payload: Event) -> FFIResult<()>;
 }
 
 pub trait TelioLoggerCb: Send + Sync + std::fmt::Debug {
     fn log(&self, log_level: TelioLogLevel, payload: String) -> FFIResult<()>;
 }
 
 pub trait TelioProtectCb: Send + Sync + RefUnwindSafe + std::fmt::Debug {
-    fn protect(&self, payload: i32) -> FFIResult<()>;
+    fn protect(&self, socket_id: i32) -> FFIResult<()>;
```
Comment (mathiaspeters) [inline src/ffi/types.rs:19]: Small request from android team

File: src/libtelio.udl [hunk 2]
```diff
@@ -399,7 +399,7 @@ callback interface TelioLoggerCb {
 
 callback interface TelioProtectCb {
     [Throws=TelioError]
-    void protect(i32 payload);
+    void protect(i32 socket_id);
```
Comment (mathiaspeters) [inline src/libtelio.udl:402]: Small request from android team

**Review by packgron — CHANGES_REQUESTED** [review]
You should update event docs

File: nat-lab/tests/uniffi/libtelio_remote.py [hunk 3]
```diff
@@ -31,6 +31,64 @@ def wrap(*args, **kwargs):
     return wrap
 
 
+# This function is meant to be a temporary way to deal with
+# structured events until natlab has been migrated to use
+# the types provided by the generated bindings
+def serialize_event(event: libtelio.Event) -> str:
```
[the remark this answers is not in the document]
Comment (mathiaspeters) [reply nat-lab/tests/uniffi/libtelio_remote.py:37]: The class isn't serializable and would need a custom serializer anyways

[the remark this answers is not in the document]
Comment (mathiaspeters) [reply nat-lab/tests/uniffi/libtelio_remote.py:38]: I did simplify it a bit though

**Comment by mathiaspeters** [comment]
> You should update event docs

Done

File: src/doc/telio_events.md [hunk 4]
```diff
@@ -1,37 +1,350 @@
 # Documentation For Describing Telio Events
 
-This module documents the events reported by the Telio library.
+This module documents the events reported by the Telio library. The events are reported strongly-typed in the supported languages. The body of the event variants are described in later sections with the top level event being defined as:
+
+<multi-code-select></multi-code-select>
```
Comment (packgron) [inline src/doc/telio_events.md:5]: I thinkg this was an overkill, it should be enough to point to Node in rust code, and people can look everything up in bindings, keeping structs here will just lead for this are to be constatntly out of sync with actual code.

So just leave textual info :sweat_smile:

**Review by packgron — DISMISSED** [review]
+2.0

File: nat-lab/tests/uniffi/libtelio_remote.py [hunk 5]
```diff
@@ -31,6 +32,34 @@ def wrap(*args, **kwargs):
...
+        event_dict["type"] = "relay"
+        body["conn_state"] = extract_value(body["conn_state"])
+    elif event.is_node():
+        event_dict["type"] = "node"
+        body["state"] = extract_value(body["state"])
+        if body["link_state"] is not None:
+            body["link_state"] = extract_value(body["link_state"])
+        body["path"] = extract_value(body["path"])
+    elif event.is_error():
+        event_dict["type"] = "error"
```
Comment (stalowyjez) [inline nat-lab/tests/uniffi/libtelio_remote.py:55]: Shouldn't it contain error level as well?

**Review by stalowyjez — COMMENTED** [review]
lgtm, +1

**Review by packgron — APPROVED** [review]

File: nat-lab/tests/uniffi/libtelio_remote.py (hunk 5 above)
Comment (mathiaspeters) [reply nat-lab/tests/uniffi/libtelio_remote.py:55]: That's very correct! I never came across an error event while implementing this so I missed that, but it's added now

Comment (mathiaspeters) [reply nat-lab/tests/uniffi/libtelio_remote.py:55]: Or maybe I didn't quite understand what you meant? It made me realize I didn't handle error level and error code in error events, but is that what you meant?

Comment (stalowyjez) [reply nat-lab/tests/uniffi/libtelio_remote.py:55]: Yeah, it was exactly about that

**Review by packgron — DISMISSED** [review]

**Review by packgron — APPROVED** [review]

**Pull request merged**